### DISCUSSION TITLE
How to get all cellPrecedents & cellDependents from one address?

### DISCUSSION THREAD

**Discussion by MartinDawson** [handsontable/hyperformula#889] [API Questions]
Hyperformula has these 2 functions:

`getCellPrecedents` & `getCellDependents`.

However they only get the immediate children/parents.

I need to get all the children/parents from an address in an efficient way.

The current hyperformula functions such as GraphBuilder.ts `processDependencies` expect all dependencies to be given up front, not just one node.

**Comment by izulin** [comment]
> Hyperformula has these 2 functions:
> 
> `getCellPrecedents` & `getCellDependents`.
> 
> However they only get the immediate children/parents.
> 
> I need to get all the children/parents from an address in an efficient way.
> 
> The current hyperformula functions such as GraphBuilder.ts `processDependencies` expect all dependencies to be given up front, not just one node.

`processDependencies` adds dependencies to the dependency graph, and it is ok to just add immediate dependencies there. Algorithms behind handling dependencies take care of dependencies of dependencies etc.

**Reply by MartinDawson** [reply]
Yes but what I am asking is how do we get all of the cell precedents starting from a single vertex/address?

Here's an example:

```js
  it('all precedents should be gotten from one vertex', () => {
    const [engine] = HyperFormula.buildFromArray({ cells: [
      [{ cellValue: 1 }, { cellValue: 2 }, { cellValue: 3 }],
      [{ cellValue: '=SUM(A1:B1)' }, { cellValue: '=SUMSQ(A1:B1)' }],
      [{ cellValue: '=A2+B2' }],
    ]})

    expect(engine.getAllCellPrecedents(adr('A3'))).toEqual([adr('A2'), adr('B2'), adr('A1'), adr('B1')])
  })
```

Currently, something like this is not possible in HF as far as I am aware?

I'm wondering how I can add this feature if it doesn't exist.

**Comment by bardek8** [comment]
That's right, currently HF supports only single-hop precedents/dependents.

In order to get them all, you can apply [BFS](https://en.wikipedia.org/wiki/Breadth-first_search)-like approach:

```
 1      AllCellPrecedents={start}
 2      let Q be an empty queue
 4      Q.enqueue(start)
 5      while Q is not empty do
 6          cell := Q.dequeue()
 7          S := getCellPrecedents(cell)
 9          for all cells c in S do:
10              if c is not in AllCellPrecedents then:
11                  insert w to AllCellPrecedents
12                  Q.enqueue(c)
```

**Reply by MartinDawson** [reply]
Thanks, that works well.

I see that `getCellPrecedents` returns `simpleCellAddress` and a `simpleCellRange`

For example this test returns true with the above BFS code:

```js
    const [engine] = HyperFormula.buildFromArray({ cells: [
      [{ cellValue: 1 }, { cellValue: 2 }],
      [{ cellValue: '=SUM(A1:B1)' }],
    ]})
    
   expect(engine.getAllCellPrecedents(adr('A2'))).toEqual([simpleCellRange(adr('A1'), adr('B1')), adr('A1'), adr('B1')])
```
Why does A2 have both a simpleCellRange precedent AND the equivalent in simpleCellAddress precedents?

I would have thought that it should do this instead:

```js
    const [engine] = HyperFormula.buildFromArray({ cells: [
      [{ cellValue: 1 }, { cellValue: 2 }],
      [{ cellValue: '=SUM(A1:B1)' }],
    ]})
    
   expect(engine.getAllCellPrecedents(adr('A2'))).toEqual([simpleCellRange(adr('A1'), adr('B1'))])
```

Instead of duplicating addresses.

Do simpleCellRanges always have their equivelent in simpleCellAddresses in this case? If so then I could just ignore simpleCellRanges from the return.

If not then I'm thinking to exclude simpleCellRanges addresses when the equivelent in simpleCellAddresses is already their.

**Reply by sequba** [reply]
> That's right, currently HF supports only single-hop precedents/dependents.
> 
> In order to get them all, you can apply [BFS](https://en.wikipedia.org/wiki/Breadth-first_search)-like approach

I created an issue (feature request) for that: #931 but it's not on our current roadmap so the BFS-like approach will be described in the docs (#934)

> Why does A2 have both a simpleCellRange precedent AND the equivalent in simpleCellAddress precedents?

[This page](https://handsontable.github.io/hyperformula/guide/key-concepts.html#handling-ranges) describes the key concepts of keeping ranges in the dependency graph. I also created an issue to explain it more accurately in the documentation: #930

**Reply by sequba** [reply]
@MartinDawson A thorough explanation of these functions can be found in the [Dependency Graph guide](https://handsontable.github.io/hyperformula/guide/dependency-graph.html) and API reference for [getCellDependents](https://handsontable.github.io/hyperformula/api/classes/hyperformula.html#getcelldependents) and [getCellPrecedents](https://handsontable.github.io/hyperformula/api/classes/hyperformula.html#getcellprecedents).
